leptonai/cloudrun/remote.py

```diff
@@ -205,42 +250,69 @@ def start_up(self):
+        # we will not handle this case, and we will simply raise an error.
+        if self.healthz():
+            # If the service is still up - don't do anything, just return.
+            return
+        try:
+            if not self.deployment_id:
+                # Never started up before, start up.
+                self._start_up()
+            else:
+                # Resume by re-running the photon
+                self._run_photon(resume=True)
+                if not self.client:
+                    self._create_client()
+        except Exception as e:
+            # close things first, and then raise the error.
+            self.close()
+            raise e
 
     def close(self):
         """
         Closes the current run.
         """
+        logger.debug(f"Remote: closing remote {self.unique_name}({self.display_name})")
+        # Implementation note: we will not raise any error if the close() operation
+        # fails, as we want to make sure that the close() operation is best-effort.
         self.client = None
-        global _refcount
-        _refcount[self.deployment_id] -= 1
-        if _refcount[self.deployment_id] == 0:
-            # Only when we get to refcount=0 do we remove the deployment and photons.
-            conn = api.workspace.current_connection()
-            if self.deployment_id is not None:
-                logger.debug(f"Remote: removing deployment {self.deployment_id}")
-                api.deployment.remove_deployment(conn, self.deployment_id)
-                self.deployment_id = None
-            if self.photon_id is not None:
-                logger.debug(f"Remote: removing photon {self.photon_id}")
-                api.photon.remove_remote(conn, self.photon_id)
-                self.photon_id = None
-            if self.path is not None:
-                logger.debug(f"Remote: removing local photon {self.path}")
-                api.photon.remove_local(self.path)
-                self.path = None
+        if self.deployment_id is not None:
+            logger.debug(f"Remote: removing deployment {self.deployment_id}")
+            api.deployment.remove_deployment(self.conn, self.deployment_id)
+            self.deployment_id = None
+        if self.photon_id is not None:
+            logger.debug(f"Remote: removing photon {self.photon_id}")
+            api.photon.remove_remote(self.conn, self.photon_id)
+            self.photon_id = None
+        if self.path is not None:
+            logger.debug(f"Remote: removing local photon {self.path}")
+            api.photon.remove_local(self.path)
+            self.path = None
+
+    @classmethod
+    def _atexit_cleanup(cls):
+        """
+        A best-effort cleanup function for the remote objects.
+
+        You should not call this function directly. It is invoked by atexit.
+        """
+        # For any remote items that are explicitly `del`ed but not garbage collected,
+        # we do a garbage collection here.
+        gc.collect()
```

Comment:
do we need to involve gc?

Reply:
Yep. This bugged me for quite a bit because there might be some objects that were del'ed, but not gc'ed yet. If we don't call gc, then if the interpreter exits, it is not guaranteed that python will actually call the __del__ and thus the close() function of these objects.

Now, strictly speaking, these "deleted but not gc'ed" objects should still be in the weakref _all_remotes, but it's kinda unpredictable what python garbage collector will do. So, a safe option is to first do a cleansweep gc.collect() first, and then for any remaining ones, we explicitly call the close() functions.